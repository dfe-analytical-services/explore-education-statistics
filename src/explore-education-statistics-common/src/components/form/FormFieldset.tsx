import classNames from 'classnames';
import React, { ReactNode } from 'react';
import ErrorMessage from '../ErrorMessage';
import FormGroup from './FormGroup';

export interface FormFieldsetProps {
  children?: ReactNode;
  error?: string;
  hint?: string;
  id: string;
  legend: ReactNode | string;
  legendSize?: 'xl' | 'l' | 'm' | 's';
  legendHidden?: boolean;
  className?: string;
}

const FormFieldset = ({
  children,
  error,
  hint,
  id,
  legend,
  legendSize = 'l',
  legendHidden = false,
  className,
}: FormFieldsetProps) => {
  return (
    <FormGroup hasError={!!error}>
      <fieldset
        aria-describedby={
          classNames({
            [`${id}-error`]: !!error,
            [`${id}-hint`]: !!hint,
          }) || undefined
        }
        className={classNames('govuk-fieldset', className)}
        id={id}
      >
        <legend
          className={classNames(
            'govuk-fieldset__legend',
            `govuk-fieldset__legend--${legendSize}`,
            {
              'govuk-visually-hidden': legendHidden,
            },
          )}
        >
          {legend}
        </legend>

        {hint && (
          <span className="govuk-hint" id={`${id}-hint`}>
            {hint}
          </span>
        )}

        {error && <ErrorMessage id={`${id}-error`}>{error}</ErrorMessage>}

        {children}
      </fieldset>
    </FormGroup>
  );
};

export default FormFieldset;
