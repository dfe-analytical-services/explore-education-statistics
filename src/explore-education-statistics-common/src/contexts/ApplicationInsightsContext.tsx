import { ApplicationInsights } from '@microsoft/applicationinsights-web';
import React, {
  createContext,
  ReactNode,
  useContext,
  useEffect,
  useState,
} from 'react';

const ApplicationInsightsContext = createContext<
  ApplicationInsights | undefined
>(undefined);

interface ApplicationInsightsContextProviderProps {
  children: ReactNode;
  instrumentationKey: string | Promise<string>;
}

export const ApplicationInsightsContextProvider = ({
  children,
  instrumentationKey,
}: ApplicationInsightsContextProviderProps) => {
  const [appInsights, setAppInsights] = useState<ApplicationInsights>();

  useEffect(() => {
    import('@common/services/applicationInsightsService').then(
      async ({ initApplicationInsights }) => {
        const applicationInsights = initApplicationInsights(
          await instrumentationKey,
        );

        setAppInsights(applicationInsights);
      },
    );
  }, [instrumentationKey]);

  return (
    <ApplicationInsightsContext.Provider value={appInsights}>
      {children}
    </ApplicationInsightsContext.Provider>
  );
};

export function useApplicationInsights(): ApplicationInsights | undefined {
  return useContext(ApplicationInsightsContext);
}
