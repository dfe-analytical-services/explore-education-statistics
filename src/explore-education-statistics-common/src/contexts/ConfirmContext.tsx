import useToggle from '@common/hooks/useToggle';
import React, {
  createContext,
  MutableRefObject,
  ReactNode,
  useContext,
  useMemo,
  useRef,
} from 'react';

export interface ConfirmContextValue {
  isConfirming: boolean;
  askConfirm(): Promise<boolean>;
  confirm(): void;
  cancel(): void;
}

export const ConfirmContext = createContext<ConfirmContextValue | undefined>(
  undefined,
);

interface Props {
  children: ReactNode | ((contextProps: ConfirmContextValue) => ReactNode);
}

export const ConfirmContextProvider = ({ children }: Props) => {
  const [isConfirming, toggleConfirming] = useToggle(false);

  const resolver: MutableRefObject<
    ((confirmed: boolean) => void) | null
  > = useRef(null);

  const value = useMemo<ConfirmContextValue>(() => {
    return {
      isConfirming,
      askConfirm() {
        toggleConfirming(true);

        return new Promise<boolean>(resolve => {
          resolver.current = resolve;
        });
      },
      confirm() {
        if (resolver.current) {
          resolver.current(true);
          toggleConfirming(false);
          resolver.current = null;
        }
      },
      cancel() {
        if (resolver.current) {
          resolver.current(false);
          toggleConfirming(false);
          resolver.current = null;
        }
      },
    };
  }, [isConfirming, toggleConfirming]);

  return (
    <ConfirmContext.Provider value={value}>
      {typeof children === 'function' ? children(value) : children}
    </ConfirmContext.Provider>
  );
};

export function useConfirmContext(): ConfirmContextValue {
  const context = useContext(ConfirmContext) as ConfirmContextValue;

  if (!context) {
    throw new Error('Must be used with a ConfirmContextProvider');
  }

  return context;
}
