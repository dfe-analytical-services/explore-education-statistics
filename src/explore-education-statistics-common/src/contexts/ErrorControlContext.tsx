import noop from 'lodash/noop';
import React, { createContext, ReactNode, useContext } from 'react';

export interface ErrorPages {
  forbidden: () => void;
}

export interface ErrorControlState {
  handleError: (error: unknown) => void;
  errorPages: ErrorPages;
}

export const ErrorControlContext = createContext<ErrorControlState>({
  handleError: noop,
  errorPages: {
    forbidden: noop,
  },
});

interface ProviderProps {
  children: ReactNode;
  value: ErrorControlState;
}

export const ErrorControlContextProvider = ({
  children,
  value,
}: ProviderProps) => {
  return (
    <ErrorControlContext.Provider value={value}>
      {children}
    </ErrorControlContext.Provider>
  );
};

export function useErrorControl(): ErrorControlState {
  return useContext(ErrorControlContext);
}
