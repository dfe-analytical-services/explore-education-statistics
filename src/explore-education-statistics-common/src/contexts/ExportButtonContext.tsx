import { createContext, useContext } from 'react';

export const ExportButtonContext =
  createContext<React.RefObject<HTMLDivElement> | null>(null);

export const useExportButtonContext = () => {
  const context = useContext(ExportButtonContext);
  if (!context) {
    throw new Error('useExportButtonContext must be used within a RefProvider');
  }
  return context;
};
