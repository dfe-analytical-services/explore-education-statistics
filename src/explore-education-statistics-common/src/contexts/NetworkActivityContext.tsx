import useMountedRef from '@common/hooks/useMountedRef';
import {
  RequestInterceptor,
  ResponseInterceptor,
} from '@common/services/api/Client';
import React, {
  createContext,
  ReactNode,
  SetStateAction,
  useCallback,
  useContext,
  useEffect,
  useState,
} from 'react';

const EVENT_NETWORK_REQUEST = 'ees:network_request';
const EVENT_NETWORK_REQUEST_ERROR = 'ees:network_request_error';
const EVENT_NETWORK_RESPONSE = 'ees:network_response';
const EVENT_NETWORK_RESPONSE_ERROR = 'ees:network_response_error';

const dataAttribute = 'data-network-activity';

export type NetworkActivityStatus = 'active' | 'idle';

export interface NetworkActivityState {
  status: NetworkActivityStatus;
  requestCount: number;
}

const NetworkActivityContext = createContext<NetworkActivityState | undefined>(
  undefined,
);

export interface NetworkActivityContextProviderProps {
  children?: ReactNode;
  idleTimeout?: number;
}

/**
 * Provider that listens for network activity events.
 *
 * Upon receiving these events, the provider will update the `data-network-activity`
 * attribute on the document body to notify other listeners (outside of React) on
 * the current state.
 *
 * This is primarily used by Robot Framework tests to identify when the network is
 * idle and the page can be considered to be 'loaded' before proceeding with assertions.
 *
 * This provider works by using client interceptors to hook into the request/response
 * lifecycle. These dispatch custom DOM events that we can listen for in the provider.
 */
export function NetworkActivityContextProvider({
  children,
  idleTimeout = 500,
}: NetworkActivityContextProviderProps) {
  const isMountedRef = useMountedRef();
  const [state, _setState] = useState<NetworkActivityState>({
    status: 'idle',
    requestCount: 0,
  });

  const setState = useCallback(
    (setter: SetStateAction<NetworkActivityState>) => {
      if (!isMountedRef.current) {
        return;
      }

      _setState(prevState => {
        const nextState =
          typeof setter === 'function' ? setter(prevState) : setter;

        document.body.setAttribute(dataAttribute, nextState.status);

        return nextState;
      });
    },
    [isMountedRef],
  );

  useEffect(() => {
    if (document.body.hasAttribute(dataAttribute)) {
      throw new Error(
        `Cannot have more than once instance of ${NetworkActivityContextProvider.name}`,
      );
    }

    document.body.setAttribute(dataAttribute, 'idle');

    return () => {
      document.body.removeAttribute(dataAttribute);
    };
  }, []);

  useEffect(() => {
    const handleRequest = () => {
      setState(prevState => {
        return {
          status: 'active',
          requestCount: prevState.requestCount + 1,
        };
      });
    };

    const handleResponse = () => {
      setState(prevState => ({
        ...prevState,
        requestCount: prevState.requestCount - 1,
      }));

      setTimeout(() => {
        setState(prevState => {
          return {
            ...prevState,
            status: prevState.requestCount === 0 ? 'idle' : 'active',
          };
        });
      }, idleTimeout);
    };

    window.addEventListener(EVENT_NETWORK_REQUEST, handleRequest);
    window.addEventListener(EVENT_NETWORK_REQUEST_ERROR, handleResponse);
    window.addEventListener(EVENT_NETWORK_RESPONSE, handleResponse);
    window.addEventListener(EVENT_NETWORK_RESPONSE_ERROR, handleResponse);

    return () => {
      window.removeEventListener(EVENT_NETWORK_REQUEST, handleRequest);
      window.removeEventListener(EVENT_NETWORK_REQUEST_ERROR, handleResponse);
      window.removeEventListener(EVENT_NETWORK_RESPONSE, handleResponse);
      window.removeEventListener(EVENT_NETWORK_RESPONSE_ERROR, handleResponse);
    };
  }, [idleTimeout, setState]);

  return (
    <NetworkActivityContext.Provider value={state}>
      {children}
    </NetworkActivityContext.Provider>
  );
}

export function useNetworkActivityContext(): NetworkActivityState {
  const context = useContext(NetworkActivityContext);

  if (!context) {
    throw new Error(
      `Must be used within a ${NetworkActivityContextProvider.name}`,
    );
  }

  return context;
}

export const networkActivityRequestInterceptor: RequestInterceptor = {
  onRequest: config => {
    if (typeof window !== 'undefined') {
      window.dispatchEvent(new CustomEvent(EVENT_NETWORK_REQUEST));
    }

    return config;
  },
  onError: error => {
    if (typeof window !== 'undefined') {
      window.dispatchEvent(new CustomEvent(EVENT_NETWORK_REQUEST_ERROR));
    }

    return Promise.reject(error);
  },
};

export const networkActivityResponseInterceptor: ResponseInterceptor = {
  onResponse: response => {
    if (typeof window !== 'undefined') {
      window.dispatchEvent(new CustomEvent(EVENT_NETWORK_RESPONSE));
    }

    return response;
  },
  onError: error => {
    if (typeof window !== 'undefined') {
      window.dispatchEvent(new CustomEvent(EVENT_NETWORK_RESPONSE_ERROR));
    }

    return Promise.reject(error);
  },
};
