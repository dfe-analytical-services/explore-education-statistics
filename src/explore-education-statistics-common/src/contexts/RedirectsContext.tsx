import { Redirects } from '@common/utils/url/applyRedirectRules';
import React, {
  createContext,
  ReactNode,
  useCallback,
  useContext,
  useMemo,
  useState,
} from 'react';

export interface RedirectsContextState {
  redirects: Redirects;
}

export const RedirectsContext = createContext<
  RedirectsContextState | undefined
>(undefined);

interface Props {
  children: ReactNode;
  refetchRedirects: () => Promise<Redirects>;
}

//let defaultRedirects = { publications: [], methodologies: [] };

export const RedirectsContextProvider = async ({
  children,
  refetchRedirects,
}: Props) => {
  const defaultRedirects = useMemo<Redirects>(
    async () => await refetchRedirects(),
    [refetchRedirects],
  );
  // const [redirectsState, setRedirectsState] = useState<RedirectsContextState>({
  //   redirects: defaultRedirects,
  // });

  // const setState = useCallback(async () => {
  //   const fetchedRedirects = await refetchRedirects();

  //   setRedirectsState({ redirects: fetchedRedirects });
  // }, [refetchRedirects]);

  return (
    <RedirectsContext.Provider value={{ redirects: defaultRedirects }}>
      {children}
    </RedirectsContext.Provider>
  );
};

export function useRedirectsContext(): RedirectsContextState {
  const context = useContext(RedirectsContext) as RedirectsContextState;

  if (!context) {
    throw new Error('Must be used with a RedirectsContextProvider');
  }

  return context;
}
