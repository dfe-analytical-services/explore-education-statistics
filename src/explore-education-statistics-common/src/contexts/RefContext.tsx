import {
  createContext,
  MutableRefObject,
  ReactNode,
  useContext,
  useRef,
} from 'react';

export const RefContext = createContext<MutableRefObject<null> | undefined>(
  undefined,
);

export const RefContextProvider = ({ children }: { children: ReactNode }) => {
  const ref = useRef(null);

  return <RefContext.Provider value={ref}>{children}</RefContext.Provider>;
};

export function useRefContext(): MutableRefObject<null> | undefined {
  const context = useContext(RefContext) as MutableRefObject<null> | undefined;

  if (!context) {
    throw new Error('Must be used within a RefContextProvider');
  }

  return context;
}
