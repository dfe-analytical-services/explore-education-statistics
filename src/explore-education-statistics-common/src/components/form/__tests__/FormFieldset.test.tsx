import { getAllDescribedBy } from '@common-test/queries';
import { FormIdContextProvider } from '@common/components/form/contexts/FormIdContext';
import { render, screen } from '@testing-library/react';
import React from 'react';
import FormFieldset from '../FormFieldset';

describe('FormFieldset', () => {
  test('renders correctly with required props', () => {
    render(
      <FormFieldset id="test-fieldset" legend="Fill the form">
        <input type="text" name="test-input-1" />
        <input type="text" name="test-input-2" />
      </FormFieldset>,
    );

    const legend = screen.getByRole('group', { name: 'Fill the form' });
    expect(legend).toBeInTheDocument();
    expect(legend.innerHTML).toMatchSnapshot();
  });

  test('renders correctly with hint', () => {
    const { container } = render(
      <FormFieldset
        id="test-fieldset"
        legend="Fill the form"
        hint="All fields are required"
      >
        <input type="text" name="test-input-1" />
        <input type="text" name="test-input-2" />
      </FormFieldset>,
    );

    expect(screen.getByText('All fields are required')).toBeInTheDocument();
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders correctly with error', () => {
    const { container } = render(
      <FormFieldset
        id="test-fieldset"
        legend="Fill the form"
        error="There was an error"
      >
        <input type="text" name="test-input-1" />
        <input type="text" name="test-input-2" />
      </FormFieldset>,
    );

    expect(screen.getByText('There was an error')).toBeInTheDocument();
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('aria-describedby is equal to the hint id', () => {
    render(
      <FormFieldset
        id="test-fieldset"
        legend="Fill the form"
        hint="All fields are required"
      />,
    );

    expect(screen.getByRole('group')).toHaveAttribute(
      'aria-describedby',
      'test-fieldset-hint',
    );
  });

  test('aria-describedby is equal to the error id', () => {
    render(
      <FormFieldset
        id="test-fieldset"
        legend="Fill the form"
        error="There was an error"
      />,
    );

    expect(screen.getByRole('group')).toHaveAttribute(
      'aria-describedby',
      'test-fieldset-error',
    );
  });

  test('aria-describedby contains both hint and error ids', () => {
    render(
      <FormFieldset
        id="test-fieldset"
        legend="Fill the form"
        hint="All fields are required"
        error="There was an error"
      />,
    );

    const group = screen.getByRole('group');
    expect(group).toHaveAttribute('id', 'test-fieldset');

    const describingElements = getAllDescribedBy(group);

    expect(describingElements).toHaveLength(2);
    expect(describingElements[0]).toHaveAttribute('id', 'test-fieldset-error');
    expect(describingElements[0]).toHaveTextContent('There was an error');

    expect(describingElements[1]).toHaveAttribute('id', 'test-fieldset-hint');
    expect(describingElements[1]).toHaveTextContent('All fields are required');
  });

  test('setting `legendSize` to xl applies class correctly', () => {
    const { container } = render(
      <FormFieldset
        id="test-fieldset"
        legend="Fill the form"
        hint="All fields are required"
        legendSize="xl"
      >
        <input type="text" name="test-input-1" />
        <input type="text" name="test-input-2" />
      </FormFieldset>,
    );

    expect(container.querySelector('legend')).toHaveClass(
      'govuk-fieldset__legend--xl',
    );
  });

  test('setting `legendHidden` to true hides the legend', () => {
    const { container } = render(
      <FormFieldset
        id="test-fieldset"
        legend="Fill the form"
        hint="All fields are required"
        legendHidden
      >
        <input type="text" name="test-input-1" />
        <input type="text" name="test-input-2" />
      </FormFieldset>,
    );

    expect(container.querySelector('legend')).toHaveClass(
      'govuk-visually-hidden',
    );
  });

  test('renders with correct ids with form context', () => {
    render(
      <FormIdContextProvider id="testForm">
        <FormFieldset
          id="test-fieldset"
          legend="Fill the form"
          hint="All fields are required"
          error="There was an error"
          legendHidden
        />
      </FormIdContextProvider>,
    );

    const group = screen.getByRole('group');
    expect(group).toHaveAttribute('id', 'testForm-test-fieldset');

    const describingElements = getAllDescribedBy(group);

    expect(describingElements).toHaveLength(2);
    expect(describingElements[0]).toHaveAttribute(
      'id',
      'testForm-test-fieldset-error',
    );
    expect(describingElements[0]).toHaveTextContent('There was an error');

    expect(describingElements[1]).toHaveAttribute(
      'id',
      'testForm-test-fieldset-hint',
    );
    expect(describingElements[1]).toHaveTextContent('All fields are required');
  });
});
