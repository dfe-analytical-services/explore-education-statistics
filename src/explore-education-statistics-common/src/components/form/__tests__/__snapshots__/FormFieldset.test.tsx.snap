// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormFieldset renders correctly with error 1`] = `
<div class="govuk-form-group govuk-form-group--error">
  <fieldset aria-describedby="test-fieldset-error"
            class="govuk-fieldset"
            id="test-fieldset"
  >
    <legend class="govuk-fieldset__legend govuk-fieldset__legend--l">
      Fill the form
    </legend>
    <span class="govuk-error-message"
          id="test-fieldset-error"
    >
      There was an error
    </span>
    <input type="text"
           name="test-input-1"
    >
    <input type="text"
           name="test-input-2"
    >
  </fieldset>
</div>
`;

exports[`FormFieldset renders correctly with hint 1`] = `
<div class="govuk-form-group">
  <fieldset aria-describedby="test-fieldset-hint"
            class="govuk-fieldset"
            id="test-fieldset"
  >
    <legend class="govuk-fieldset__legend govuk-fieldset__legend--l">
      Fill the form
    </legend>
    <span class="govuk-hint"
          id="test-fieldset-hint"
    >
      All fields are required
    </span>
    <input type="text"
           name="test-input-1"
    >
    <input type="text"
           name="test-input-2"
    >
  </fieldset>
</div>
`;

exports[`FormFieldset renders correctly with required props 1`] = `
<div class="govuk-form-group">
  <fieldset class="govuk-fieldset"
            id="test-fieldset"
  >
    <legend class="govuk-fieldset__legend govuk-fieldset__legend--l">
      Fill the form
    </legend>
    <input type="text"
           name="test-input-1"
    >
    <input type="text"
           name="test-input-2"
    >
  </fieldset>
</div>
`;
